server/src/test/java/keywhiz/service/daos/SecretContentDAOMockTest.java

```diff
@@ -0,0 +1,119 @@
+  private static final Map<String, String> NO_METADATA = null;
+  private static final long NO_EXPIRY = 0;
+  private static final long NO_NOW = 0;
+
+  @Rule
+  public ExpectedException thrown = ExpectedException.none();
+
+  @Test
+  public void rejectsSecretWithSizeGreaterThanMaximumSecretSize() {
+    Long maximumSecretSizeInBytes = 100L;
+    String secret = StringUtils.repeat("x", (int) (maximumSecretSizeInBytes + 1));
+
+    SecretContentDAO dao = createDaoWithMaximumSecretSize(maximumSecretSizeInBytes);
+
+    thrown.expect(IllegalArgumentException.class);
+
+    createSecret(dao, secret);
+  }
+
+  @Test
+  public void acceptsSecretWithSizeEqualToMaximumSize() {
+    Long maximumSecretSizeInBytes = 100L;
+    String secret = StringUtils.repeat("x", (int) (long) (maximumSecretSizeInBytes));
+
+    SecretContentDAO dao = createDaoWithMaximumSecretSize(maximumSecretSizeInBytes);
+
+    createSecret(dao, secret);
+  }
+
+  @Test
+  public void acceptsSecretWithSizeLessThanMaximumSize() {
+    Long maximumSecretSizeInBytes = 100L;
+    String secret = StringUtils.repeat("x", (int) (long) (maximumSecretSizeInBytes));
+
+    SecretContentDAO dao = createDaoWithMaximumSecretSize(maximumSecretSizeInBytes);
+
+    createSecret(dao, secret);
+  }
+
+  @Test
+  public void acceptsLargeSecretIfNoMaximumSizeSpecified() {
+    // 65KB
+    String secret = StringUtils.repeat("x", 65 * (int) Math.pow(2, 10));
+
+    SecretContentDAO dao = createDaoWithMaximumSecretSize(null);
+
+    createSecret(dao, secret);
+  }
+
+  private void createSecret(SecretContentDAO dao, String secret) {
+    dao.createSecretContent(
+        NO_SECRET_ID,
+        secret,
+        NO_HMAC,
+        NO_CREATOR,
+        NO_METADATA,
+        NO_EXPIRY,
+        NO_NOW);
+  }
+
+  private SecretContentDAO createDaoWithMaximumSecretSize(Long maximumSecretSizeInBytes) {
+    KeywhizConfig config = new KeywhizConfig();
+    config.setMaximumSecretSizeInBytes(maximumSecretSizeInBytes);
+
+    MockConnection mockConnection = new MockConnection(new DataProvider());
+    DSLContext context = DSL.using(mockConnection);
+
+    return new SecretContentDAO(
+        context,
+        new ObjectMapper(),
+        Mockito.mock(SecretContentMapper.class),
+        Mockito.mock(RowHmacGenerator.class),
+        config);
+  }
+
+  private static class DataProvider implements MockDataProvider {
+    @Override public MockResult[] execute(MockExecuteContext mockExecuteContext) {
+      DSLContext context = DSL.using(SQLDialect.MYSQL);
+
+      List<MockResult> results = new ArrayList<>();
+
+      for (int i = 0; i < mockExecuteContext.autoGeneratedKeys(); i++) {
+        Record1<Long> record = context.newRecord(SecretsContent.SECRETS_CONTENT.ID);
+        Record1<Long> values = record.values(1L);
+        MockResult result = new MockResult(values);
+        results.add(result);
+      }
+
+      return results.toArray(new MockResult[0]);
+    }
+  }
```

Comment:
This is taken from the Jooq unit testing docs. Unfortunately there is no actual implementation of MockDataProvider so you always have to roll your own. The behavior we see here is to satisfy the case where an insert statement is run and an auto-generated ID is expected as the result. In other words, the behavior of SecretsContentRecord.store() from our DAO.

The alternative is many, many layers of mocking the Jooq APIs which I very much wanted to avoid.

Reply:
Dang, making us roll our own method! Out of curiosity, did you know to roll this out (rather than mock Jooq APIs) immediately, or did you try to mock Jooq APIs and realize it was going to be tedious?

Reply:
I started with mocking first, but Jooq's API is builder-ish in that it likes to return lots of objects, and those objects are built up by lots of calls. So, multiple methods on multiple classes would have been necessary for mocking. I've been down that road before and wanted to avoid it if at all possible. OTOH it's a one-time investment, but hopefully so is a working example of how to test "by the rules".